app/src/main/java/tech/relaycorp/gateway/domain/publicsync/DeliverPublicParcels.kt

```diff
@@ -0,0 +1,85 @@
+package tech.relaycorp.gateway.domain.publicsync
+
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.flatMapLatest
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.take
+import tech.relaycorp.gateway.common.Logging.logger
+import tech.relaycorp.gateway.data.database.StoredParcelDao
+import tech.relaycorp.gateway.data.disk.DiskMessageOperations
+import tech.relaycorp.gateway.data.disk.MessageDataNotFoundException
+import tech.relaycorp.gateway.data.model.RecipientLocation
+import tech.relaycorp.gateway.data.model.StoredParcel
+import tech.relaycorp.gateway.domain.DeleteParcel
+import tech.relaycorp.gateway.domain.LocalConfig
+import tech.relaycorp.gateway.pdc.PoWebClientBuilder
+import tech.relaycorp.poweb.PoWebClient
+import tech.relaycorp.relaynet.bindings.pdc.ClientBindingException
+import tech.relaycorp.relaynet.bindings.pdc.RejectedParcelException
+import tech.relaycorp.relaynet.bindings.pdc.ServerException
+import tech.relaycorp.relaynet.bindings.pdc.Signer
+import tech.relaycorp.relaynet.cogrpc.readBytesAndClose
+import java.util.logging.Level
+import javax.inject.Inject
+
+@JvmSuppressWildcards
+class DeliverPublicParcels
+@Inject constructor(
+    private val storedParcelDao: StoredParcelDao,
+    private val diskMessageOperations: DiskMessageOperations,
+    private val poWebClientBuilder: PoWebClientBuilder,
+    private val localConfig: LocalConfig,
+    private val deleteParcel: DeleteParcel
+) {
+
+    suspend fun deliver(keepAlive: Boolean) {
+        logger.info("Delivering parcels to Public Gateway (keepAlive=$keepAlive)")
+
+        val poWebClient = poWebClientBuilder.build()
+        val parcelsQuery =
+            storedParcelDao.observeForRecipientLocation(RecipientLocation.ExternalGateway)
+        val parcelsFlow =
+            if (keepAlive) {
+                // one at a time
+                parcelsQuery.filter { it.any() }.map { it.first() }
+            } else {
+                // just one batch, but one at a time
+                parcelsQuery.take(1).flatMapLatest { it.asFlow() }
+            }
+
+        parcelsFlow.collect { parcel ->
+            try {
+                deliverParcel(poWebClient, parcel)
+            } catch (e: ServerException) {
+                logger.log(Level.INFO, "Could not deliver parcels due to server error", e)
+            }
+        }
+    }
+
+    @Throws(ServerException::class)
+    private suspend fun deliverParcel(poWebClient: PoWebClient, parcel: StoredParcel) {
+        val parcelStream = parcel.getInputStream() ?: return
+
+        try {
+            poWebClient.deliverParcel(parcelStream.readBytesAndClose(), getSigner())
+            deleteParcel.delete(parcel)
+        } catch (e: RejectedParcelException) {
+            logger.log(Level.WARNING, "Could not deliver rejected parcel", e)
```

Comment:
BTW, the parcel should be deleted in this case. The server is basically saying that it will never accept it.